pinot-server/src/main/java/com/linkedin/pinot/server/starter/helix/HelixServerStarter.java

```diff
@@ -177,9 +186,97 @@ public HelixServerStarter(String helixClusterName, String zkServer, Configuratio
     serverMetrics.addCallbackGauge("memory.allocationFailureCount", PinotDataBuffer::getAllocationFailureCount);
   }
 
-  private void updateInstanceConfigInHelix(int adminApiPort, boolean shuttingDown) {
+  private void waitForAllSegmentsLoaded() {
+    if (_helixServerConfig.getBoolean(CommonConstants.Server.CONFIG_OF_STARTER_ENABLE_SEGMENTS_LOADING_CHECK, CommonConstants.Server.DEFAULT_STARTER_ENABLE_SEGMENTS_LOADING_CHECK)) {
+      long startTime = System.currentTimeMillis();
+      int serverStarterTimeout = _helixServerConfig.getInt(CommonConstants.Server.CONFIG_OF_STARTER_TIMEOUT_IN_SECONDS, CommonConstants.Server.DEFAULT_STARTER_TIMEOUT_IN_SECONDS);
+      long endTime = TimeUnit.SECONDS.toMillis(startTime + serverStarterTimeout);
+      boolean allSegmentsLoaded = false;
+      while (System.currentTimeMillis() < endTime) {
+        long timeToSleep = Math.min(TimeUnit.MILLISECONDS.toSeconds(endTime - System.currentTimeMillis()), 10 /* Sleep 10 seconds as default*/);
+        if (ServiceStatus.getServiceStatus() == Status.GOOD) {
+          LOGGER.info("All the segments are fully loaded into Pinot server, time taken: {} seconds", TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis() - startTime));
+          allSegmentsLoaded = true;
+          break;
+        }
+        try {
+          int numSegmentsLoaded = getNumSegmentLoaded();
+          int numSegmentsToLoad = getNumSegmentsToLoad();
+          LOGGER.warn("Waiting for all segments to be loaded, current progress: [ {} / {} ], sleep {} seconds...", numSegmentsLoaded, numSegmentsToLoad, timeToSleep);
+          logSegmentsLoadingInfo();
```

Comment:
It might be good to log this only at the end when we are bailing out - else it might result in too much logging.

Reply:
Only logging per table level progress, when full segments bootstrap is not completed on time.

Reply:
Thanks for fixing it!